src/Model_ELBDM/ELBDM_GetTimeStep_Velocity.cpp

```diff
@@ -0,0 +1,152 @@
+//
+// Parameter   :  lv : Target refinement level
+//
+// Return      :  MaxdS_dx
+//-------------------------------------------------------------------------------------------------------
+real GetMaxVelocity( const int lv, bool ExcludeWaveCells )
+{
+   // Maximum velocity for calculation of time step zero for wave patches
+   // since we are only concerned with a velocity dependent time step criterion for the fluid solver
+   if ( amr->use_wave_flag[lv] )
+      return 0;
+
+
+   const bool IntPhase_No       = false;
+   const bool DE_Consistency_No = false;
+   const real MinDens_No        = -1.0;
+   const real MinPres_No        = -1.0;
+   const real MinTemp_No        = -1.0;
+   const real MinEntr_No        = -1.0;
+   const int  NGhost            = 1;               // number of ghost zones to calculate dS_dx
+   const int  Size_Flu          = PS2 + 2*NGhost;  // size of the array Flu_Array
+   const int  NPG               = 1;               // number of patch groups (must be ONE here)
+   const int  NCOMP1            = 1;               // we only need to retrieve phase for velocity criterion
+   real (*Flu_Array)[NCOMP1][Size_Flu][Size_Flu][Size_Flu] = NULL;
+   real (*Wave_Patch_Array)[NCOMP1][Size_Flu][Size_Flu][Size_Flu] = NULL;
+
+   real dS_dx, MaxdS_dx, _dh, _dh2, GradS[3];
+   int im, ip, jm, jp, km, kp, I, J, K;
+
+   MaxdS_dx = 0.0;
+   _dh      = (real)1.0/amr->dh[lv];
+   _dh2     = (real)0.5*_dh;
+
+// construct global tree structure
+   LB_GlobalTree GlobalTree;
+
+#  pragma omp parallel private( Flu_Array, dS_dx, GradS, \
+                                im, ip, jm, jp, km, kp, I, J, K)
+   {
+      Flu_Array = new real [NPG][NCOMP1][Size_Flu][Size_Flu][Size_Flu];
+
+//    loop over all patches
+#     pragma omp for reduction( max:MaxdS_dx ) schedule( runtime )
+      for (int PID0=0; PID0<amr->NPatchComma[lv][1]; PID0+=NPG*8)
+      {
+
+//       prepare phase with NGhost ghost zone on each side (any interpolation scheme can be used)
+         Prepare_PatchData( lv, Time[lv], &Flu_Array[0][0][0][0][0], NULL, NGhost, NPG, &PID0, _PHAS, _NONE,
+                            INT_CQUAD, INT_NONE, UNIT_PATCHGROUP, NSIDE_06, IntPhase_No, OPT__BC_FLU, BC_POT_NONE,
+                            MinDens_No, MinPres_No, MinTemp_No, MinEntr_No, DE_Consistency_No );
+
+
+
+//       evaluate dS_dt
+         for (int k=NGhost; k<Size_Flu-NGhost; k++)    {  km = k - 1;    kp = k + 1;   K = k - NGhost;
+         for (int j=NGhost; j<Size_Flu-NGhost; j++)    {  jm = j - 1;    jp = j + 1;   J = j - NGhost;
+         for (int i=NGhost; i<Size_Flu-NGhost; i++)    {  im = i - 1;    ip = i + 1;   I = i - NGhost;
+
+//       skip velocities of cells that have a wave counterpart on the refined levels
+
+         long GID0                   = GlobalTree.PID2GID(PID0, lv);
+         bool DoNotCalculateVelocity = false;
+         if ( ExcludeWaveCells )
+         {
+            for (int LocalID=0; LocalID<8; LocalID++ )
```

Comment:
This loop is a bit confusing since in principle we only need to check a single patch for which `(I, J, K)` belongs to. So consider adding a comment that `(I, J, K)` outside the patch `GID0 + Local` will be skipped by `ELBDM_HasWaveCounterpart()` automatically.